l1-contracts/deploy-scripts/PrepareZKChainRegistrationCalldata.s.sol

```diff
@@ -0,0 +1,358 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+// solhint-disable no-console
+
+import {Script, console2 as console} from "forge-std/Script.sol";
+import {stdToml} from "forge-std/StdToml.sol";
+
+import {IBridgehub} from "contracts/bridgehub/IBridgehub.sol";
+import {Bridgehub} from "contracts/bridgehub/Bridgehub.sol";
+import {L2ContractHelper} from "contracts/common/libraries/L2ContractHelper.sol";
+import {AddressAliasHelper} from "contracts/vendor/AddressAliasHelper.sol";
+import {L1SharedBridge} from "contracts/bridge/L1SharedBridge.sol";
+import {IStateTransitionManager} from "contracts/state-transition/IStateTransitionManager.sol";
+import {IGovernance} from "contracts/governance/IGovernance.sol";
+import {IERC20} from "forge-std/interfaces/IERC20.sol";
+import {Utils} from "./Utils.sol";
+
+/**
+ * @title Prepare ZKChain Registration Calldata
+ * @dev Script to prepare calldata for registering a new ZKChain on the bridgehub
+ *
+ * This script prepares calldata for registering a new ZKChain on the bridgehub
+ * That includes 3 steps:
+ * 1. Register base token on the bridgehub
+ * 2. Register the new ZKChain on the bridgehub
+ * 3. Initialize the L2 bridge on the L1 shared bridge
+ *
+ * The script precomputes the address of the L2 bridge to generate the calldata for the third step.
+ * It assumes that L1 governance is the owner of the L2 bridge and the L2 bridge is deployed by the msg.sender of the script.
+ *
+ * The script reads the configuration from a TOML file `script-config/prepare-registration-calldata.toml`
+ * and writes the output to a TOML file `script-out/output-prepare-registration-calldata.toml`.
+ *
+ * The output contains 4 fields:
+ * - scheduleCalldataStageOne: calldata for scheduling the first stage
+ * - executeCalldataStageOne: calldata for executing the first stage
+ * - scheduleCalldataStageTwo: calldata for scheduling the second stage
+ * - executeCalldataStageTwo: calldata for executing the second stage
+ * (stage 2 of the execution is the registration of the L2 bridge on the L1 shared bridge)
+ *
+ * The separation is required to ensure that there is no period of time where the L2 bridge is registered, so users
+ * can send their funds there, but they will be burned in case L2 bridge is not initialized by the chain operator.
+ * It is meant to be executed only after the L2 bridge is deployed.
+ *
+ * How to use:
+ * 1. Create a TOML file `script-config/prepare-registration-calldata.toml`, reference config at `deploy-script-config-template/config-prepare-registration-calldata.toml`.
+ * 2. Run the script impersonating the address that will deploy the L2 bridge, `forge script --rpc-url $RPC --sender $DEPLOYER PrepareZKChainRegistrationCalldata.s.sol`
```

Comment:
how do I know which address will deploy the L2 bridge? 
or any address is ok?

Reply:
so DEPLOYER doesn't need any special parameters on L2 - great. Lets mention it here.
(especially as step4 is deploying this bridge right?)

Reply:
mentioned